axes/handlers/database.py

```diff
@@ -119,58 +119,48 @@ def user_login_failed(
         # 2. database query: Calculate the current maximum failure number from the existing attempts
         failures_since_start = 1 + self.get_failures(request, credentials)
 
-        # 3. database query: Insert or update access records with the new failure data
-        try:
-            attempt = AccessAttempt.objects.get(
+        if not (settings.AXES_ONLY_USER_FAILURES and username is None):
+            # 3. database query: Insert or update access records with the new failure data
+            attempt, created = AccessAttempt.objects.get_or_create(
```

Comment:
Using `get_or_create` with `save` will invoke database 3 times in a case where a user does not exist: 

1. first a query to check if it exists, 
2. then a creation if it does not, and
3. finally a save round to update the counters and metadata.

The old logic only uses two database invocations at most, and this MR potentially increases the database query load by 50% per attempt for cases where an access attempt does not previously exist?

- https://docs.djangoproject.com/en/3.2/ref/models/querysets/#get-or-create
- https://github.com/django/django/blob/ca9872905559026af82000e46cde6f7dedc897b6/django/db/models/query.py#L571

The proposed logic does avoid a race condition in case where the [database is not configured for atomic requests with `ATOMIC_REQUESTS = True`](https://docs.djangoproject.com/en/3.2/ref/settings/#std:setting-DATABASE-ATOMIC_REQUESTS) on the engine level AND the unique fields are defined on the model, but this comes at a cost here. Can we define the unique fields in the current configuration, because users can configure the lockout lookup fields with the following settings:

- AXES_ONLY_USER_FAILURES
- AXES_LOCK_OUT_BY_COMBINATION_USER_AND_IP
- AXES_LOCK_OUT_BY_USER_OR_IP
- AXES_USE_USER_AGENT

Reply:
I think the additional save() could be prevented in case of "created"

Marking the whole function as @transaction.atomic might be an easier option, good point 

I do not think that the database configuration has an impact with the current implementation.

Reply:
The transaction atomic should not be required anymore, and now also understanding the rather complex query inside get_failures, I would rather stick to the local solution.

Reply:
OK 👍